### PR TITLE
rama policy support restrict cpu size by reference control knob

### PR DESCRIPTION
**Pull request by luomingmeng**
#### What type of PR is this?

Enhancements
#### What this PR does / why we need it:
rama policy support restrict cpu size by reference control knob.

### CODE REVIEW

File: cmd/katalyst-agent/app/options/sysadvisor/qosaware/resource/cpu/provision/policy_rama.go [hunk 1]
```diff
@@ -19,10 +19,14 @@ package provision
 import (
 	"github.com/spf13/pflag"
 
+	"github.com/kubewharf/katalyst-core/pkg/agent/sysadvisor/types"
 	provisionconfig "github.com/kubewharf/katalyst-core/pkg/config/agent/sysadvisor/qosaware/resource/cpu/provision"
 )
 
 type PolicyRamaOptions struct {
+	AdjustBasedPolicyName           string
```
Comment (waynepeking348) [inline cmd/katalyst-agent/app/options/sysadvisor/qosaware/resource/cpu/provision/policy_rama.go:27]: extra to another struct

File: pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/advisor.go [hunk 2]
```diff
@@ -61,8 +61,8 @@ const (
 )
 
 func init() {
-	provisionpolicy.RegisterInitializer(types.CPUProvisionPolicyCanonical, provisionpolicy.NewPolicyCanonical)
-	provisionpolicy.RegisterInitializer(types.CPUProvisionPolicyRama, provisionpolicy.NewPolicyRama)
+	provisionpolicy.RegisterInitializer(types.CPUProvisionPolicyCanonical, 0, provisionpolicy.NewPolicyCanonical)
```
Comment (waynepeking348) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/advisor.go:64]: const

File: pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/provisionpolicy/policy_rama.go [hunk 3]
```diff
@@ -105,34 +105,85 @@ func (p *PolicyRama) Update() error {
...
+	if err != nil {
+		return err
 	}
 
 	p.regulator.Regulate(cpuAdjustedRestricted)
 
 	return nil
 }
 
+func (p *PolicyRama) cpuAdjustedBasedControlKnob(cpuAdjustedRaw float64) (float64, error) {
```
Comment (waynepeking348) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/provisionpolicy/policy_rama.go:125]: `do something by something else`

File: pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/provisionpolicy/policy_rama.go [hunk 4]
```diff
@@ -105,34 +105,85 @@ func (p *PolicyRama) Update() error {
...
+	}
+
+	if cpuAdjustedRestricted != cpuAdjustedRaw {
+		klog.Infof("[qosaware-cpu-rama] restrict cpu adjusted from %.2f to %.2f, reason: %v", cpuAdjustedRaw, cpuAdjustedRestricted, reason)
+	}
+
+	return cpuAdjustedRestricted, nil
+}
+
+func (p *PolicyRama) cpuAdjustedBasedReclaimedUsage(cpuAdjustedRaw float64) (float64, error) {
```
Comment (waynepeking348) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/provisionpolicy/policy_rama.go:159]: as previously

File: pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/provisionpolicy/policy_rama.go [hunk 5]
```diff
@@ -105,34 +105,85 @@ func (p *PolicyRama) Update() error {
...
+	if p.conf.PolicyRama.AdjustBasedPolicyName == "" {
+		return cpuAdjustedRaw, nil
+	}
+
+	basedControlKnob, ok := p.ControlEssentials.BasedControlKnobs[p.conf.PolicyRama.AdjustBasedPolicyName]
+	if !ok {
+		return 0, fmt.Errorf("get control knob of adjust based policy %v failed", p.conf.PolicyRama.AdjustBasedPolicyName)
+	}
+
+	basedNonReclaimedCPUSize := basedControlKnob[types.ControlKnobNonReclaimedCPUSize].Value
```
Comment (waynepeking348) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/provisionpolicy/policy_rama.go:135]: not to assume that `types.ControlKnobNonReclaimedCPUSize` must exist

File: pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/provisionpolicy/policy_rama.go [hunk 6]
```diff
@@ -105,34 +105,85 @@ func (p *PolicyRama) Update() error {
...
+
+	leftBoundary := math.Min(basedNonReclaimedCPUSize-p.conf.PolicyRama.AdjustBasedPolicyMaxGap, basedNonReclaimedCPUSize*(1-p.conf.PolicyRama.AdjustBasedPolicyMaxGapRatio))
```
Comment (waynepeking348) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/provisionpolicy/policy_rama.go:137]: min/max

File: pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/region_base.go [hunk 7]
```diff
@@ -90,6 +91,7 @@ type QoSRegionBase struct {
 	// the former has higher priority; provisionPolicyNameInUse indicates the provision
 	// policy in-use currently
 	provisionPolicies        []*internalProvisionPolicy
+	provisionUpdatePolicies  []*internalProvisionPolicy
```
Comment (sun-yuliang) [inline pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/region_base.go:94]: The naming is confusing. provisionPoliciesInUpdatePriority might be better. Add comment for it.

**Review by sun-yuliang — APPROVED** [review]

**Pull request closed without merging**